tests/src/Functional/RequestTest.php

```diff
@@ -0,0 +1,982 @@
+   * code as simple as possible, and hence not require them to specify the
+   * 'http_errors = FALSE' request option, nor do we want them to have to
+   * convert Drupal Url objects to strings.
+   *
+   * We also don't want to follow redirects automatically, to ensure these tests
+   * are able to detect when redirects are added or removed.
+   *
+   * @see \GuzzleHttp\ClientInterface::request()
+   *
+   * @param string $method
+   *   HTTP method.
+   * @param \Drupal\Core\Url $url
+   *   URL to request.
+   * @param array $request_options
+   *   Request options to apply.
+   *
+   * @return \Psr\Http\Message\ResponseInterface
+   */
+  protected function request($method, Url $url, array $request_options) {
+    $request_options[RequestOptions::HTTP_ERRORS] = FALSE;
+    $request_options[RequestOptions::ALLOW_REDIRECTS] = FALSE;
+    $request_options = $this->decorateWithXdebugCookie($request_options);
+    $client = $this->getSession()->getDriver()->getClient()->getClient();
+    return $client->request($method, $url->setAbsolute(TRUE)->toString(), $request_options);
+  }
+
+  /**
+   * Adds the Xdebug cookie to the request options.
+   *
+   * @param array $request_options
+   *   The request options.
+   *
+   * @return array
+   *   Request options updated with the Xdebug cookie if present.
+   */
+  protected function decorateWithXdebugCookie(array $request_options) {
+    $session = $this->getSession();
+    $driver = $session->getDriver();
+    if ($driver instanceof BrowserKitDriver) {
+      $client = $driver->getClient();
+      foreach ($client->getCookieJar()->all() as $cookie) {
+        if (isset($request_options[RequestOptions::HEADERS]['Cookie'])) {
+          $request_options[RequestOptions::HEADERS]['Cookie'] .= '; ' . $cookie->getName() . '=' . $cookie->getValue();
+        }
+        else {
+          $request_options[RequestOptions::HEADERS]['Cookie'] = $cookie->getName() . '=' . $cookie->getValue();
+        }
+      }
+    }
+    return $request_options;
+  }
+
+  /**
+   * Check the expected response.
+   *
+   * @param \Psr\Http\Message\ResponseInterface $response
+   *   The response from the test request.
+   * @param string $entity_type_id
+   *   Then entity type.
+   * @param string|null $bundle_name
+   *   The bundle name or NULL.
+   */
+  protected function checkExpectResponse(ResponseInterface $response, $entity_type_id, $bundle_name = NULL) {
+    $this->assertEquals('200', $response->getStatusCode());
+    $key = $entity_type_id . ($bundle_name ? ":$bundle_name" : '');
+    $expected_responses = $this->getExpectResponses();
+    if (isset($expected_responses[$key])) {
+      $this->assertEquals($expected_responses[$key], json_decode($response->getBody()->getContents(), TRUE));
+    }
+  }
+
+  /**
+   * Gets expected responses.
+   *
+   * @return array
+   *   The expected responses.
+   */
+  protected function getExpectResponses() {
```

Comment:
Added json files to /schemata/tests/expectations

Thought they shouldn't go under /tests/src and also they could be used at some point for kernel based tests.